tutorials-v5/time-evolution/019_optomechanical-steadystate.md

```diff
@@ -0,0 +1,274 @@
+            "use_precond": True,
+            "atol": 1e-15,
+            "maxiter": int(1e5),
+            **precond_options,
+        }
+        use_rcm = True
+    else:
+        solver_options = {}
+        use_rcm = False
+
+    start = time.time()
+    rho_ss = steadystate(
+        H,
+        c_ops,
+        method=method,
+        solver=solver,
+        use_rcm=use_rcm,
+        **solver_options,
+    )
+    end = time.time()
+
+    print(f"Solver: {solver}, Time: {np.round(end-start, 5)}")
+    rho_mech = ptrace(rho_ss, 1)
+    mech_dms.append(rho_mech)
+
+rho_mech = mech_dms[0]
+mech_dms = [mech_dm.data.as_ndarray() for mech_dm in mech_dms]
+```
+
+### Check Consistency of Solutions
+
+
+Can check to see if the solutions are the same by looking at the number of nonzero elements (NNZ) in the difference between mechanical density matrices.
+
+```python
+for kk in range(len(mech_dms)):
+    c = np.where(
+            np.abs(mech_dms[kk].flatten() - mech_dms[0].flatten()) > 1e-5
+        )[0]
+    print(f"#NNZ for k = {kk} : {len(c)}")
+```
+
+## Plot the Mechanical Oscillator Wigner Function
+
+
+It is known that the density matrix for the mechanical oscillator is diagonal in the Fock basis due to phase diffusion. If we look at the `hinton()` plot of the density matrix, we can see the magnitude of the diagonal elements is higher, such that the non-diagonal have a vanishing importance.
+
+```python
+hinton(rho_mech.data.as_ndarray(), x_basis=[""] * Nm, y_basis=[""] * Nm);
+```
+
+However some small off-diagonal terms show up during the factorization process, which we can display by the using `plt.spy()`.
+
+```python
+plt.spy(rho_mech.data.as_ndarray(), markersize=1);
+```
+
+Therefore, to remove this error, let use explicitly take the diagonal elements and form a new operator out of them.
+
+```python
+diag = rho_mech.diag()
+rho_mech2 = qdiags(diag, 0, dims=rho_mech.dims, shape=rho_mech.shape)
+hinton(rho_mech2, x_basis=[""] * Nm, y_basis=[""] * Nm);
+```
+
+Now lets compute the oscillator Wigner function and plot it to see if there are any regions of negativity.
+
+```python
+xvec = np.linspace(-20, 20, 256)
+W = wigner(rho_mech2, xvec, xvec)
+wmap = wigner_cmap(W, shift=-1e-5)
+```
+
+```python
+fig, ax = plt.subplots(figsize=(8, 6))
+c = ax.contourf(xvec, xvec, W, 256, cmap=wmap)
+ax.set_xlim([-10, 10])
+ax.set_ylim([-10, 10])
+plt.colorbar(c, ax=ax);
+```
+
+## About
+
+```python
+about()
+```
+
+### Testing
+
+```python
+# assert obtained steady-state via mesolve evolution
+psi0 = tensor(basis(Nc), basis(Nm))
+rho0 = psi0 @ psi0.dag()
+tlist = np.linspace(0, 1500, 1500)
+rho_evolve = mesolve(H, rho0, tlist, c_ops)
+rho_final = ptrace(rho_evolve.states[-1], 1)
+assert fidelity(rho_mech, rho_final) > 0.99
+
+# assert magnitude of diagonal elements are more than non-diagonal
+rho_ss_dmat = rho_ss.data.to_array()
+rho_ss_diag = np.diag(rho_ss.diag())
+assert np.all(np.abs(rho_ss_diag) - np.abs(rho_ss_dmat) <= 0)
```

Comment:
This seems true for any matrix -- i.e. that `abs(just the diagonals) - abs(full matrix)` has entries `<= 0`. Let's remove this block and think of a better check to add instead.

Reply:
Nice catch, I changed this test to now properly assert diagonal dominance.